### PR TITLE
Follow Docker Token Authentication Specification

### PR DESCRIPTION
**Pull request by joshspicer**
ref: https://github.com/devcontainers/cli/issues/322

Implements the [Docker Registry v2 authentication via central service specification](https://docs.docker.com/registry/spec/auth/token/) to improve compatibility with registry services.  While not officially part of the distribution specification, this specification is widely used by various container registries.  This change has been tested with Azure Container Registry (acr) and GitHub Container Registry (ghcr). 

As outlined in the spec, requests will initially be made to a registry without any authentication.  If authentication is required by the server, a `401` response will be set and the `WWW-Authenticate` header will be set with information on how to exchange for a bearer token for the given resource, as well as which scopes to request.  

```
realm="https://ghcr.io/token",service="ghcr.io",scope="repository:devcontainers/features:pull,push"
```

Registries like GHCR and ACR require this token exchange even for anonymous access (pulling a public artifact).

This PR wraps all pulling and pushing HTTP operations in `requestEnsureAuthenticated(...)`, which will read the `WWW-Authenticate` header on `401` and attempt to negotiate a token with the server.  The last token to be successfully used will be cached and first attempted on subsequent requests.

This implementation supports exchanging for a Bearer token (as described in the linked specification from Docker), as well as setting the authentication header with Basic auth credentials (as expected by the `registry/registry` reference implementation

Additionally, this PR will now attempt to read the `$HOME/.docker/config.json` file on disk (if one exists) and will share these with the registry server to get a more privileged Bearer token.  By following [the 'Authenticate using token' steps on the Azure Container Registry Docs](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-repository-scoped-permissions#authenticate-using-token), one will be able to both publish and pull from an ACR that requires authentication. The existing `DEVCONTAINERS_OCI_AUTH` environment variable is still supported, and is preferred over reading the home folder config.json.  For `GHCR`, the environment's `GITHUB_TOKEN` is always preferred.

A test has been added to validate that anonymously pulling a Feature from ACR works end to end.  That Feature was pushed to ACR via the changes in this PR.

image

### CODE REVIEW

**Comment by joshspicer** [comment]
cc/ @brendandburns

File: src/spec-configuration/containerCollectionsOCI.ts [hunk 1]
```diff
@@ -3,19 +3,19 @@ import * as semver from 'semver';
...
 export const DEVCONTAINER_MANIFEST_MEDIATYPE = 'application/vnd.devcontainers';
 export const DEVCONTAINER_TAR_LAYER_MEDIATYPE = 'application/vnd.devcontainers.layer.v1+tar';
 export const DEVCONTAINER_COLLECTION_LAYER_MEDIATYPE = 'application/vnd.devcontainers.collection.layer.v1+json';
 
-export type HEADERS = { 'authorization'?: string; 'user-agent': string; 'content-type'?: string; 'accept'?: string; 'content-length'?: string };
 
 export interface CommonParams {
 	env: NodeJS.ProcessEnv;
 	output: Log;
+	cachedAuthHeader?: string;
```
Comment (chrmarti) [inline src/spec-configuration/containerCollectionsOCI.ts:18]: Should this be a map from registries to auth headers?

File: src/spec-configuration/httpOCIRegistry.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+			return Buffer.from(userToken)
+				.toString('base64');
+		}
+	} else {
+		const homeDir = os.homedir();
+		if (homeDir) {
+			const dockerConfigPath = path.join(homeDir, '.docker', 'config.json');
+			if (await isLocalFile(dockerConfigPath)) {
+				try {
+					const dockerConfig: DockerConfigFile = JSON.parse((await readLocalFile(dockerConfigPath)).toString());
```
Comment (chrmarti) [inline src/spec-configuration/httpOCIRegistry.ts:148]: Could use `jsonc.parse` for additional robustness. (Though Docker might actually be strict from what I have seen.)

File: src/spec-configuration/httpOCIRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+					if (dockerConfig.auths && dockerConfig.auths[registry] && dockerConfig.auths[registry].auth) {
```
Comment (chrmarti) [inline src/spec-configuration/httpOCIRegistry.ts:150]: We should also support Docker credential helpers, these are more common in local setups. I recently figured out how to use them for the Dev Containers extension. The code will need to be adapted, it implements a credential helper proxy: https://github.com/microsoft/vscode-remote-containers/blob/main/src/common/dockerCredentialHelper.ts

Ok to do in a separate PR.

**Review by chrmarti — CHANGES_REQUESTED** [review]
Great! Left a few comments.

File: src/spec-configuration/containerCollectionsOCI.ts (hunk 1 above)
Comment (joshspicer) [reply src/spec-configuration/containerCollectionsOCI.ts:18]: Good point. For the push operations I think this would be fine (we push to one registry at a time), but for pulling perhaps this `params` object is reused? Nonetheless makes sense to make it a map to be sure we never send an auth header to the wrong server.

File: src/spec-configuration/httpOCIRegistry.ts (hunk 3 above)
Comment (joshspicer) [reply src/spec-configuration/httpOCIRegistry.ts:150]: Thanks for the pointer.  I have a TODO comment inline here to add credential helper support, and will tackle that in a subsequent PR.

Comment (joshspicer) [reply src/spec-configuration/httpOCIRegistry.ts:150]: https://github.com/devcontainers/cli/issues/349

**Review by alexdima — APPROVED** [review]
